File: android/src/main/java/org/ligi/gobandroid_hd/ui/gnugo/GnuGoHelper.java
package org.ligi.gobandroid_hd.ui.gnugo;

import android.content.Context;
import android.content.Intent;
import org.ligi.axt.AXT;
import org.ligi.gobandroid_hd.logic.CellImpl;
import org.ligi.gobandroid_hd.logic.GoGame;
import org.ligi.gobandroid_hd.logic.StatefulGoBoard;

public class GnuGoHelper {

    public final static String INTENT_ACTION_NAME = "org.ligi.gobandroidhd.ai.gnugo.GnuGoService";

    public static boolean isGnuGoAvail(Context ctx) {
        return (AXT.at(new Intent(INTENT_ACTION_NAME)).isServiceAvailable(ctx.getPackageManager(), 0));
    }


    public static boolean checkGnuGoSync(final String board_str, final GoGame game) {
        final StatefulGoBoard b = new StatefulGoBoard(game.getStatelessGoBoard());
        final String[] split_board = board_str.split("\n");

        for (int gnugo_y = 2; gnugo_y <= b.getSize() + 1; gnugo_y++) {
            final String act_line = split_board[gnugo_y].replace(" ", "").replace("" + (game.getBoardSize() - (gnugo_y - 2)), "");
            for (int gnugo_x = 0; gnugo_x < b.getSize(); gnugo_x++) {
                final CellImpl cell = new CellImpl(gnugo_x, gnugo_y - 2);

                if (act_line.length() < (gnugo_x + 1) ||
                    (act_line.charAt(gnugo_x) == '.' && !game.getVisualBoard().isCellFree(cell)) ||
                    (act_line.charAt(gnugo_x) == 'X' && !game.getVisualBoard().isCellBlack(cell)) ||
                    (act_line.charAt(gnugo_x) == 'O' && !game.getVisualBoard().isCellWhite(cell))) {
                    return false;
                }

            }

        }

        return true;
    }
}

File: android/src/main/java/org/ligi/gobandroid_hd/ui/gnugo/GnuGoHelper.kt
package org.ligi.gobandroid_hd.ui.gnugo

import android.content.Context
import android.content.Intent
import org.ligi.gobandroid_hd.logic.CellImpl
import org.ligi.gobandroid_hd.logic.GoGame
import org.ligi.gobandroid_hd.logic.StatefulGoBoard
import org.ligi.kaxt.isServiceAvailable

object GnuGoHelper {

    val INTENT_ACTION_NAME = "org.ligi.gobandroidhd.ai.gnugo.GnuGoService"

    fun isGnuGoAvail(ctx: Context)= Intent(INTENT_ACTION_NAME).isServiceAvailable(ctx.packageManager, 0)

    fun checkGnuGoSync(board_str: String, game: GoGame): Boolean {
        val b = StatefulGoBoard(game.statelessGoBoard)
        val split_board = board_str.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        for (gnugo_y in 2..b.size + 1) {
            val act_line = split_board[gnugo_y].replace(" ", "").replace("" + (game.boardSize - (gnugo_y - 2)), "")
            for (gnugo_x in 0..b.size - 1) {
                val cell = CellImpl(gnugo_x, gnugo_y - 2)

                if (act_line.length < gnugo_x + 1 ||
                        act_line[gnugo_x] == '.' && !game.visualBoard.isCellFree(cell) ||
                        act_line[gnugo_x] == 'X' && !game.visualBoard.isCellBlack(cell) ||
                        act_line[gnugo_x] == 'O' && !game.visualBoard.isCellWhite(cell)) {
                    return false
                }

            }

        }

        return true
    }
}
